components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/main/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImpl.java

```diff
@@ -964,99 +1276,220 @@ private Optional<String> getRoleIdFromAudience(String roleName, String audienceT
-                                                  String unsharingInitiatedOrgId)
-            throws ResourceSharingPolicyMgtException {
+    /**
+     * Retrieves the role IDs currently assigned to a shared user within an organization.
+     *
+     * @param userAssociation The user association object containing user and organization details.
+     * @return A list of role IDs currently assigned to the shared user.
+     */
+    private List<String> getCurrentSharedRoleIdsForSharedUser(UserAssociation userAssociation)
+            throws OrganizationManagementException, IdentityRoleManagementException {
 
-        getResourceSharingPolicyHandlerService().deleteResourceSharingPolicyInOrgByResourceTypeAndId(
-                organizationId, ResourceType.USER, associatedUserId, unsharingInitiatedOrgId);
+        String userId = userAssociation.getUserId();
+        String orgId = userAssociation.getOrganizationId();
+        String tenantDomain = getOrganizationManager().resolveTenantDomain(orgId);
+
+        List<String> allUserRolesOfSharedUser = getRoleManagementService().getRoleIdListOfUser(userId, tenantDomain);
+
+        return getOrganizationUserSharingService().getSharedUserRolesFromUserRoles(allUserRolesOfSharedUser,
+                tenantDomain);
     }
 
-    private String getOrganizationName(String organizationId) throws OrganizationManagementException {
+    /**
+     * Shares a user with a specified organization and assigns roles if present.
+     * This is where the user association will be created, and roles will be assigned to the shared user if any
+     * roles are present.
+     * It attempts to share the user with the specified organization and assigns roles accordingly.
+     *
+     * @param orgId                 The ID of the organization to share the user with.
+     * @param baseUserShare         The base user share object containing user and role information.
+     * @param sharingInitiatedOrgId The ID of the organization that initiated the sharing.
+     */
+    private void shareAndAssignRolesIfPresent(String orgId, BaseUserShare baseUserShare,
+                                              String sharingInitiatedOrgId)
+            throws OrganizationManagementException, IdentityRoleManagementException, UserSharingMgtException {
 
-        return getOrganizationManager().getOrganizationNameById(organizationId);
+        String associatedUserId = baseUserShare.getUserId();
+        List<String> roleIds = baseUserShare.getRoles();
+        UserAssociation userAssociation;
+
+        try {
+            userAssociation = shareUserWithOrganization(orgId, associatedUserId, sharingInitiatedOrgId);
+        } catch (OrganizationManagementException e) {
+            String errorMessage = String.format(ERROR_CODE_USER_SHARE.getMessage(), associatedUserId, e.getMessage());
+            LOG.error(errorMessage, e);
+            return;
+        }
+
+        //Assign roles if any are present.
```

Comment:
```suggestion
        // Assign roles if any are present.
```
Update formatting of all these type comments

Reply:
Addressed in this [PR](https://github.com/wso2-extensions/identity-organization-management/pull/470).